walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,187 @@
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    // If an event's attendee list contains a names of an attendee that needs to attend
+    // (from Collection<String> attendees, then we add the event's TimeRange to a list.
+    events.forEach(event -> {
+      Collection<String> eventAttendants = event.getAttendees();
+
+      for (String attendee : eventAttendants) {
+        if (attendees.contains(attendee)) {
+          TimeRange currentEventTimeRange = event.getWhen();
+          
+          // If the currentEventTimeRange isn't merged into a TimeRange within the
+          // unavailableTimeRanges list, then add the currentEventTimeRange to the list.
+          if (!mergeOnOverlap(unavailableTimeRanges, currentEventTimeRange)) {
+            unavailableTimeRanges.add(currentEventTimeRange);
+          }
+
+          break;
+        }
+      }
+    });
+
+    return unavailableTimeRanges;
+  }
+
+  private boolean mergeOnOverlap(Collection<TimeRange>timeRanges, TimeRange newTimeRange) {
+    // The default value is set to false, which is converted to true if a merge takes place.
+    boolean overlaps = false;
+
+    // Store items that we will add and remove from the collection after we finish iterating
+    // through it in order to avoid ConcurrentModificationException from being thrown
+    List<TimeRange> toRemove = new ArrayList<>();
+    List<TimeRange> toAdd = new ArrayList<>();
+
+    for (TimeRange takenTimeRange : timeRanges) {
+      if (newTimeRange.overlaps(takenTimeRange)) {
+        TimeRange combinedTimeRange = mergeTimeRanges(newTimeRange, takenTimeRange);
+        toRemove.add(takenTimeRange);
+        toAdd.add(combinedTimeRange);
+        overlaps = true;
+      }
+    }
+
+    for (TimeRange timeRange : toRemove) {
+      timeRanges.remove(timeRange);
+    }
+
+    for (TimeRange timeRange : toAdd) {
+      timeRanges.add(timeRange);
+    }
+
+    return overlaps;
+  }
+
+  private TimeRange mergeTimeRanges(TimeRange firstTimeRange, TimeRange secondTimeRange) {
+    TimeRange mergedTimeRange = TimeRange.fromStartEnd(
+        Math.min(firstTimeRange.start(), secondTimeRange.start()),
+        Math.max(firstTimeRange.end(), secondTimeRange.end()),
+        false
+    );
+
+    return mergedTimeRange;
+  }
+
+  private List<TimeRange> getAvailableTimeRanges(List<TimeRange> unavailableTimeRanges, long duration) {
+    if (unavailableTimeRanges.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> availableTimeRanges = new ArrayList<>();
+
+    // Check for available time starting at the beginning of the day.
+    if (unavailableTimeRanges.get(0).start() >= duration) {
+      availableTimeRanges.add(TimeRange.fromStartEnd(
+          TimeRange.WHOLE_DAY.start(),
+          unavailableTimeRanges.get(0).start(),
+          false
+      ));
+    }
+    
+    int i;
```

Comment:
Do we need to declare it here? Why don't we declare it in the for loop?

I don't think it has to be referenced outside the for loop.

Reply:
Done.